lambdas/access_counts/index.py

```diff
@@ -190,20 +267,65 @@ def delete_temp_dir():
         if errors:
             print(errors)
-            raise Exception("Failed to delete the temporary directory")
+            raise Exception(f"Failed to delete dir: bucket={bucket!r}, prefix={prefix!r}")
 
 
 def handler(event, context):
-    delete_temp_dir()
-
-    # Drop old Athen tables from previous runs.
+    # End of the CloutTrail time range we're going to look at. Subtract 5min because CloudTrail takes some time to log the events.
+    end_ts = datetime.utcnow() - timedelta(minutes=5)
+
+    # Start of the CloudTrail time range: the end timestamp from the previous run, or a year ago if it's the first run.
+    try:
+        start_ts = datetime.utcfromtimestamp(float(s3.get_object(Bucket=QUERY_RESULT_BUCKET, Key=LAST_UPDATE_KEY)['Body'].read()))
+    except s3.exceptions.NoSuchKey as ex:
+        start_ts = end_ts - timedelta(days=365)
+
+    start_date = start_ts.date()
+    end_date = end_ts.date()
+
+    # Delete the temporary directory where Athena query results are written to.
+    delete_dir(QUERY_RESULT_BUCKET, QUERY_TEMP_DIR)
+
+    # Create a CloudTrail table, but only with partitions for the last N days, to avoid scanning all of the data.
+    # A bucket can have data for multiple accounts and multiple regions, so those need to be handled first.
+    partition_queries = []
+    for account_response in s3.list_objects_v2(Bucket=CLOUDTRAIL_BUCKET, Prefix='AWSLogs/', Delimiter='/').get('CommonPrefixes') or []:
+        account = account_response['Prefix'].split('/')[1]
+        for region_response in s3.list_objects_v2(Bucket=CLOUDTRAIL_BUCKET, Prefix=f'AWSLogs/{account}/CloudTrail/', Delimiter='/').get('CommonPrefixes') or []:
+            region = region_response['Prefix'].split('/')[3]
+            date = start_date
+            while date <= end_date:
+                query = ADD_CLOUDTRAIL_PARTITION.format(
+                    account=account,
+                    region=region,
+                    year=date.year,
+                    month=date.month,
+                    day=date.day
+                )
+                partition_queries.append(query)
+                date += timedelta(days=1)
+
+    # Drop old Athena tables from previous runs.
     # (They're in the DB owned by the stack, so safe to do.)
-    for query_id in [run_query(DROP_OBJECT_ACCESS_LOG), run_query(DROP_PACKAGE_HASHES)]:
+    for query_id in [run_query(DROP_CLOUDTRAIL), run_query(DROP_OBJECT_ACCESS_LOG), run_query(DROP_PACKAGE_HASHES)]:
         wait_for_query(query_id)
 
-    for query_id in [run_query(CREATE_OBJECT_ACCESS_LOG), run_query(CREATE_PACKAGE_HASHES)]:
+    # Create new Athena tables.
+    for query_id in [run_query(CREATE_CLOUDTRAIL), run_query(CREATE_OBJECT_ACCESS_LOG), run_query(CREATE_PACKAGE_HASHES)]:
         wait_for_query(query_id)
 
+    # Create CloudTrail partitions, after the CloudTrail table is created.
+    for query_id in [run_query(q) for q in partition_queries]:
+        wait_for_query(query_id)
```

Comment:
Because these queries are issued sequentially (right?) we shouldn't hit the service limits of a max of 20 DDL simultaneous DDL queries? In any case do these rate limits deserves a note somewhere https://docs.aws.amazon.com/athena/latest/ug/service-limits.html?

Reply:
Actually, they're issued in parallel (hence the weird `for` loop with a list comprehension, rather than `wait_for_query(run_query(q))`). I tested it on 365 days, where it would've run 365 queries in parallel, but nothing failed... But you're right, we shouldn't do that.

Reply:
Since we use Glue now (true?) I think [these are the right docs](https://docs.aws.amazon.com/general/latest/gr/aws_service_limits.html#limits_glue). Same numbers

My reading on "Number of DDL queries you can submit at the same time" is not how many queries are running, but how many you can submit in a single request. I could be totally wrong. Maybe the safe thing to do is: only set 20 partitions in per `run_query` and space out the calls.